src/Microsoft.Diagnostics.DebugServices.Implementation/ContextService.cs

```diff
@@ -0,0 +1,227 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Diagnostics.DebugServices.Implementation
+{
+    /// <summary>
+    /// Manages the current target, thread and runtime contexts
+    /// </summary>
+    public class ContextService : IContextService
+    {
+        protected readonly IHost Host;
+        private ITarget _currentTarget;
+        private IThread _currentThread;
+        private IRuntime _currentRuntime;
+
+        public readonly ServiceProvider ServiceProvider;
+
+        public ContextService(IHost host)
+        {
+            Host = host;
+
+            ServiceProvider = new ServiceProvider(new Func<IServiceProvider>[] {
+                // First check the current runtime for the service
+                () => GetCurrentRuntime()?.Services,
+                // If there is no target, then provide just the global services
+                () => GetCurrentTarget()?.Services ?? host.Services
+            });
+
+            // These services depend on no caching
+            ServiceProvider.AddServiceFactoryWithNoCaching<ITarget>(GetCurrentTarget);
```

Comment:
In the future, I plan on more reflection based DI for services that would injection the dependences via the constructor or properties.  This PR moved a little in that direction by just depending on a IServiceProvider instead of specifically the global, per-target or per-context.  There is still a ways to go.  

The hosts and implementations are currently responsibility for setup/initializing the services. Eventually there will be services like SOSHost, ClrMDHelper, etc. that will be completely separate from the host and will both make themselves available (export) and receive the dependent services via some kind of DI.